### PR TITLE
Add debug asserts to AOTAutograd for input consistency with compilation

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89710
* __->__ #89702
* #89701

Fixes https://github.com/pytorch/torchdynamo/issues/1927

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -1457,7 +1499,36 @@ def compiled_function(*args):
...
 
-    return compiled_function
+    if not config.debug_assert:
+        return compiled_function
+
+    flat_requires_grad = [a.requires_grad if isinstance(a, Tensor) else None for a in flat_args]
+
+    @wraps(compiled_function)
+    def debug_compiled_function(*args):
+        # TODO: Check aliasing relationships
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1509]: yes :)

File: functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -1457,7 +1499,36 @@ def compiled_function(*args):
...
+        # TODO: Check strides for metadata mutation
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1510]: Checking my understanding - is this saying that we should check that the strides of inputs are exactly the same across runs? That seems at odds with our goal of eventually being stride agnostic.

File: functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -1097,9 +1102,7 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Tensor], aot_config: AOTConfi
         add_dupe_map[i] = j
         j += 1
 
-    # Fastpath
-    if not dropped_args:
-        return aot_dispatch_deduplicated_autograd(flat_fn, flat_args, aot_config)
+    unique_args = j
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1105]: nit: `num_unique_args`?

File: functorch/_src/aot_autograd.py [hunk 4]
```diff
@@ -1109,6 +1112,45 @@ def remove_dupe_args(args):
...
+                f"runtime there were only {len(seen)} distinct arguments.  " +
+                GUARD_BUG_BOILERPLATE
+            )
+            """
+            return f(*args)
+
+        return debug_wrapper
+
+    # Fastpath
+    if not dropped_args:
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1151]: Doesn't that mean that when we have no dupe args, we'll always skip our debug checks about requires-gradness (and eventually input alias info?)

File: functorch/_src/aot_autograd.py [hunk 5]
```diff
@@ -1457,7 +1499,36 @@ def compiled_function(*args):
...
+    def debug_compiled_function(*args):
+        # TODO: Check aliasing relationships
+        # TODO: Check strides for metadata mutation
+        # (NB: ideally, this logic is factored out of this function and
+        # you move these debug checks there)
+
+        # Check requires grad.  Bad case is when we compiled with
+        # requires_grad = False, but input requires_grad = True
+        # (vice versa is OK; we compute a gradient and then throw
+        # it away when it hits the input.)
```
Comment (bdhirsh) [inline functorch/_src/aot_autograd.py:1517]: This is definitely reasonable, but it feels like a compilation policy decision? Maybe we should include this in a section of compilation policy (when we add one / doesn't need to be in this PR).

There's the degenerate case where a user runs a function once with a bunch of requires_grad=True inputs, and then multiple times with many of the inputs set to requires_grad=False. Then we're compiling a backward function that does a bunch of compute for no reason. Although I can see this case not mattering much in practice, and compiling fewer times being more important.

**Review by bdhirsh — APPROVED** [review]
nice!

File: functorch/_src/aot_autograd.py (hunk 2 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1510]: What I mean here, is if we make assumptions about what the input strides of the tensor should be, we should test those assumptions are upheld. This may not be an exact test; in dynamic shapes, this would correspond to testing that the symbolic formulas match up, but in static shapes it would be an exact test. In any case, the point is to avoid accuracy errors from unchecked `as_strided` invocation

File: functorch/_src/aot_autograd.py (hunk 5 above)
Comment (ezyang) [reply functorch/_src/aot_autograd.py:1517]: There are two parts: is it sound to use a requires_grad = True graph when requires_grad = False, and then the policy decision.

The soundness argument is unconditionally true. It's just like how we have derivative formulas that say how to calculate derivatives for all inputs, but you may not actually use all of them. The assertion here only cares about whether or not something is sound: it needs to assert when something is unsound (not when you've made a suboptimal policy decision.)

You've accuracy described the policy problem. But in fact, the "compile more, avoid dead compute" policy IS, hypothetically, what is currently implemented. This is because Dynamo installs guards on all tensors which require `requires_grad` to match exactly. However, in an earlier CI run, I tried the strong invariant, and it actually fails in practice. So there is some fishy going on with the guards, which may warrant some further exploration. But maybe for now we should just log a warning when this happens.
